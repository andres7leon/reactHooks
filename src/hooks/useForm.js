import { useEffect, useState } from "react"

export const useForm = (initialState = {}) => {

    const [values, setValues] = useState(initialState);

    // const { name } = values;


    const reset = () => {
        setValues( initialState );
    }

    const handleInputChange = ({target}) => {
        setValues({
            ...values,
            [ target.name ]: target.value
        })
    }
    
    // useEffect(() => {
    //     console.log('entra efectos')
    // })

    return [ values, handleInputChange, reset ]
}
