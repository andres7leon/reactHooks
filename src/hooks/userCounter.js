import { useState } from "react"

export const UserCounter = ( initialState = 10 ) => {

    const [state, setState] = useState(initialState);


    const incremet = ( factor = 1) => {
        setState( state + factor);
    }

    const decremet = ( factor = 1) => {
        setState( state - factor);
    }

    const reset = () => {
        setState(initialState);
    }

    return {
        state,
        incremet,
        decremet,
        reset
    }
}
