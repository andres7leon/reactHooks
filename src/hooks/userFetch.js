import { useEffect, useRef, useState } from "react"

export const UserFetch = ( url ) => {

    const isMounted = useRef(true);
    const [state, setState] = useState( { data: null, loading: true, error: null} );

    useEffect(() => {
    
        return () => {
            isMounted.current = false;
        }
    }, [])

    useEffect(() => {
        fetch( url )
            .then( res => res.json())
            .then( data => {
                setTimeout( ()=> {
                    if ( isMounted.current ) {
                        setState({
                            loading: false,
                            error: null,
                            data
                        })
                    }
                }, 3500)
            })
            .catch( () => {
                setState({
                    loading: false,
                    error: 'No se pudo cargar la info',
                    data: null
                });
            })
    }, [url])

    return state;
}
