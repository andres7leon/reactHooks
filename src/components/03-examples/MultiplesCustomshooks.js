import React from 'react'
import { UserCounter } from '../../hooks/userCounter';
import { UserFetch } from '../../hooks/userFetch'

export const MultiplesCustomshooks = () => {

    const {state:counter, incremet} = UserCounter(1);
    const {loading, data} = UserFetch( `https://www.breakingbadapi.com/api/quotes/${counter}` );

    const { author, quote } = !!data && data[0]

    return (
        <div>
            <h3> BreakingBad </h3>
            <hr/>

            {
                loading 
                ? 
                ( 
                    <div className='alert alert-info text-center'>
                        loading...
                    </div>
                )
                :
                (
                    <blockquote className='blockquote text-right'>
                        <p className='mb-0'>{ quote }</p>
                        <small>{ author}</small>
                    </blockquote>
                )
            }

            <button className='btn btn-info' onClick={ () => incremet() }> Siguiente Frase </button>

        </div>
    )
}
