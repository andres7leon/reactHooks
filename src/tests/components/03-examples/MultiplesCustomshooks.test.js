import React from 'react';
import {shallow} from 'enzyme'
import { MultiplesCustomshooks } from '../../../components/03-examples/MultiplesCustomshooks';
import { UserFetch } from '../../../hooks/userFetch';
import { UserCounter } from '../../../hooks/userCounter';

jest.mock( '../../../hooks/userFetch' );
jest.mock( '../../../hooks/userCounter' );


describe(' Pruebas en hook MultiplesCustomshooks', () => {

    beforeEach( () =>  {
        UserCounter.mockReturnValue({
            counter: 10,
            increment: () => {}
        });
 
    });

    test('Debe de mostrarse correctamente', () => {
       
        UserFetch.mockReturnValue({
            data: null,
            loading: true,
            error: null
        });

        const wrapper = shallow( <MultiplesCustomshooks />);

        expect( wrapper ).toMatchSnapshot();
        
    });

    test('Debe de de mostrar la informacion', () => {
       
        UserFetch.mockReturnValue({
            data: [
                {
                    author: 'Andres',
                    quote: 'seeeeeee'
                }
            ],
            loading: false,
            error: null
        });

        const wrapper = shallow( <MultiplesCustomshooks />);

        expect( wrapper.find('small').text().trim() ).toBe('Andres');
        expect( wrapper.find('p').text().trim() ).toBe('seeeeeee');
        
    });
    
    
});
