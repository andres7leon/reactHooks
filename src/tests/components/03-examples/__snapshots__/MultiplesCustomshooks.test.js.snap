// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` Pruebas en hook MultiplesCustomshooks Debe de mostrarse correctamente 1`] = `
<div>
  <h3>
     BreakingBad 
  </h3>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    loading...
  </div>
  <button
    className="btn btn-info"
    onClick={[Function]}
  >
     Siguiente Frase 
  </button>
</div>
`;
